File: tools/benchmark/BgzfBench.java
import htsjdk.samtools.util.BlockCompressedInputStream;
import htsjdk.samtools.util.BlockCompressedOutputStream;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.security.MessageDigest;

/**
 * The reference side of the I/O floor benchmark: htsjdk's BGZF writer and reader, timed on the
 * same payload the Rust side uses, in the same container.
 *
 * Prints one `name=value` line per measurement so the runner can read it without parsing prose,
 * and an md5 of every stream it writes so the comparison is byte equality rather than "both
 * finished". A speed number for a path whose bytes were never compared is not a measurement of
 * this port.
 */
public class BgzfBench {
  /** The same 64-bit LCG the conformance harnesses use, so the payload is one both sides agree on. */
  static byte[] lcg(int n, long seed, int shift) {
    byte[] b = new byte[n];
    long s = seed;
    for (int i = 0; i < n; i++) {
      s = s * 6364136223846793005L + 1442695040888963407L;
      b[i] = (byte) (s >>> shift);
    }
    return b;
  }

  /** Repetitive, SAM-shaped text: the compressible end of what a real BAM's blocks look like. */
  static byte[] text(int n) {
    byte[] p = "ACGTNacgtn\tSAMrecord\tRG:Z:rg1\n".getBytes();
    byte[] b = new byte[n];
    for (int i = 0; i < n; i++) b[i] = p[i % p.length];
    return b;
  }

  static String hex(byte[] b) {
    StringBuilder s = new StringBuilder();
    for (byte x : b) s.append(String.format("%02x", x));
    return s.toString();
  }

  static String md5(byte[] b, int len) throws Exception {
    MessageDigest md = MessageDigest.getInstance("MD5");
    md.update(b, 0, len);
    return hex(md.digest());
  }

  static byte[] deflate(byte[] input, int level) throws Exception {
    ByteArrayOutputStream sink = new ByteArrayOutputStream(input.length);
    BlockCompressedOutputStream out = new BlockCompressedOutputStream(sink, (File) null, level);
    out.write(input);
    out.close();
    return sink.toByteArray();
  }

  static byte[] inflate(byte[] framed) throws Exception {
    BlockCompressedInputStream in = new BlockCompressedInputStream(new ByteArrayInputStream(framed));
    ByteArrayOutputStream sink = new ByteArrayOutputStream();
    byte[] buf = new byte[65536];
    int n;
    while ((n = in.read(buf)) > 0) sink.write(buf, 0, n);
    in.close();
    return sink.toByteArray();
  }

  public static void main(String[] args) throws Exception {
    int megabytes = args.length > 0 ? Integer.parseInt(args[0]) : 64;
    int reps = args.length > 1 ? Integer.parseInt(args[1]) : 3;
    int size = megabytes * 1024 * 1024;

    byte[][] payloads = {text(size), lcg(size, 12345L, 58)};
    String[] names = {"text", "lcg"};

    for (int p = 0; p < payloads.length; p++) {
      byte[] input = payloads[p];
      System.out.printf("payload_%s_md5=%s%n", names[p], md5(input, input.length));
      for (int level : new int[] {1, 5, 6, 9}) {
        byte[] framed = null;
        // The first run is thrown away: it pays the JIT, and the port has no equivalent cost.
        deflate(input, level);
        for (int r = 0; r < reps; r++) {
          long t0 = System.nanoTime();
          framed = deflate(input, level);
          long ns = System.nanoTime() - t0;
          System.out.printf(
              "java_deflate_%s_level%d_run%d_mbps=%.2f%n",
              names[p], level, r, (megabytes * 1e9) / ns);
        }
        System.out.printf(
            "java_deflate_%s_level%d_bytes=%d md5=%s%n",
            names[p], level, framed.length, md5(framed, framed.length));

        inflate(framed);
        for (int r = 0; r < reps; r++) {
          long t0 = System.nanoTime();
          byte[] back = inflate(framed);
          long ns = System.nanoTime() - t0;
          if (back.length != input.length) throw new IllegalStateException("round trip lost bytes");
          System.out.printf(
              "java_inflate_%s_level%d_run%d_mbps=%.2f%n",
              names[p], level, r, (megabytes * 1e9) / ns);
        }
      }
    }
  }
}

File: crates/htsjdk-bgzf/src/bin/bgzf-bench.rs
//! The port's side of the I/O floor benchmark: the same payloads, the same levels, the same
//! `name=value` lines as `tools/benchmark/BgzfBench.java`.
//!
//! It prints the digest of every framed stream it writes, so the runner compares bytes before it
//! compares seconds. Issue #78's rule is that a path with no golden gets no benchmark row; the BGZF
//! bytes have one (the `bgzf` and `zlib` suites), and this program re-establishes it on the very
//! payload it is timing rather than trusting that.

use std::io::Write;

use htsjdk_bgzf::{decompress_all, BgzfWriter};
use md5::{Digest, Md5};

fn hex(bytes: &[u8]) -> String {
    bytes.iter().map(|b| format!("{b:02x}")).collect()
}

fn md5(bytes: &[u8]) -> String {
    let mut hasher = Md5::new();
    hasher.update(bytes);
    hex(&hasher.finalize())
}

/// The same 64-bit LCG the conformance harnesses use.
fn lcg(n: usize, seed: u64, shift: u32) -> Vec<u8> {
    let mut out = vec![0u8; n];
    let mut s = seed;
    for byte in out.iter_mut() {
        s = s
            .wrapping_mul(6_364_136_223_846_793_005)
            .wrapping_add(1_442_695_040_888_963_407);
        *byte = (s >> shift) as u8;
    }
    out
}

fn text(n: usize) -> Vec<u8> {
    let pattern = b"ACGTNacgtn\tSAMrecord\tRG:Z:rg1\n";
    (0..n).map(|i| pattern[i % pattern.len()]).collect()
}

fn deflate(input: &[u8], level: u32) -> Vec<u8> {
    let mut writer = BgzfWriter::with_level(Vec::with_capacity(input.len()), level);
    writer
        .write_all(input)
        .expect("in-memory writes never fail");
    writer.into_inner().expect("in-memory writes never fail")
}

fn main() {
    let mut args = std::env::args().skip(1);
    let megabytes: usize = args
        .next()
        .map(|a| a.parse().expect("megabytes"))
        .unwrap_or(64);
    let reps: usize = args.next().map(|a| a.parse().expect("reps")).unwrap_or(3);
    let size = megabytes * 1024 * 1024;

    for (name, input) in [("text", text(size)), ("lcg", lcg(size, 12345, 58))] {
        println!("payload_{name}_md5={}", md5(&input));
        for level in [1u32, 5, 6, 9] {
            // One untimed run each way, so both sides pay their warm-up before the clock starts.
            let framed = deflate(&input, level);
            for run in 0..reps {
                let start = std::time::Instant::now();
                let framed = deflate(&input, level);
                let seconds = start.elapsed().as_secs_f64();
                println!(
                    "rust_deflate_{name}_level{level}_run{run}_mbps={:.2}",
                    megabytes as f64 / seconds
                );
                std::hint::black_box(framed);
            }
            println!(
                "rust_deflate_{name}_level{level}_bytes={} md5={}",
                framed.len(),
                md5(&framed)
            );

            let _ = decompress_all(&framed).expect("the port reads what it wrote");
            for run in 0..reps {
                let start = std::time::Instant::now();
                let back = decompress_all(&framed).expect("the port reads what it wrote");
                let seconds = start.elapsed().as_secs_f64();
                assert_eq!(back.len(), input.len(), "round trip lost bytes");
                println!(
                    "rust_inflate_{name}_level{level}_run{run}_mbps={:.2}",
                    megabytes as f64 / seconds
                );
            }
        }
    }
}
